### DISCUSSION TITLE
API Cache doesn't respect custom headers

### DISCUSSION THREAD

**Discussion by piotr-cz** [directus/directus#24885] [Feature Requests]
### Describe the Bug

Directus API caching mechanism (particularly the [`getCacheKey`](https://github.com/directus/directus/blob/v9.21.2/api/src/utils/get-cache-key.ts#L6) function) doesn't respect custom headers.

In effect when cache is enabled API returns exact same response to two requests with varying request header.

Possible solution: Add new Configuration Option to specify custom headers which will respected when creating cache key.

Related PR: https://github.com/directus/directus/pull/15537

### To Reproduce

1. Enable cache (`CACHE_ENABLED=true`)
1. Create [Custom API hook](https://docs.directus.io/extensions/hooks.html) extension with following content:
   ```ts
   import type { Application } from 'express';
   import { defineHook } from '@directus/extensions-sdk';

   export default defineHook(({ init }, { exceptions }) => {
   	const { ServiceUnavailableException } = exceptions;

   	// Note: Using middlewares.before event bypasses cache, but also bypasses CORS middleware for error response
   	init('middlewares.after', ({ app }: Record<'app', Application>) => {
   		app.use((req, _res, next) => {
   			const tokenHeaderValue = req.get('X-token');

   			if (tokenHeaderValue === 'valid') {
   				return next();
   			}

   			return next(new ServiceUnavailableException('Invalid token'));
   		});
   	});
   });
   ```

1. Send valid request (responds with `Status: 200 OK`)
   ```http
   GET http://localhost:8055/server/info HTTP/1.1
   X-token: valid
   
   Status: 200 OK
   ```

1. Send invalid request (should respond with `Status: 503 Invalid token`)
   ```http
   GET http://localhost:8055/server/info HTTP/1.1
   X-token: invalid
   
   Status: 200 OK
   ```

### Errors Shown

No errors shown

### What version of Directus are you using?

9.21.2

### What version of Node.js are you using?

16.18.0

### What database are you using?

MySQL

### What browser are you using?

Chrome

### How are you deploying Directus?

locally

**Comment by piotr-cz** [comment]
Same applies to query strings which are being sanitized (only whitelisted keys are respected)

**Comment by rijkvanzanten** [comment]
Linear: [ENG-114](https://linear.app/directus/issue/ENG-114/api-cache-doesnt-respect-custom-headers)
